components/net/decoder.rs

```diff
@@ -162,166 +167,192 @@ impl Decoder {
+        Err(e) => Poll::Ready(Some(Err(e.into()))),
     }
 }
 
 impl Stream for Gzip {
-    type Item = Chunk;
-    type Error = Error;
+    type Item = Result<Bytes, Error>;
 
-    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-        poll_with_read(&mut self.inner, &mut self.buf)
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut buf = self.buf.clone();
+        if let Ok(mut reader) = self.reader.lock() {
+            reader.waker = Some(cx.waker().clone());
+        }
+        poll_with_read(&mut self.inner, &mut buf)
     }
 }
 
 /// A brotli decoder that reads from a `brotli::Decompressor` into a `BytesMut` and emits the results
-/// as a `Chunk`.
+/// as a `Bytes`.
 struct Brotli {
     inner: Box<Decompressor<Peeked<ReadableChunks<Body>>>>,
     buf: BytesMut,
+    reader: Arc<Mutex<ReadableChunks<Body>>>,
 }
 
 impl Brotli {
     fn new(stream: ReadableChunks<Body>) -> Self {
+        let stream = Arc::new(Mutex::new(stream));
+        let reader = stream.clone();
         Self {
             buf: BytesMut::with_capacity(INIT_BUFFER_SIZE),
             inner: Box::new(Decompressor::new(Peeked::new(stream), BUF_SIZE)),
+            reader: reader,
         }
     }
 }
 
 impl Stream for Brotli {
-    type Item = Chunk;
-    type Error = Error;
+    type Item = Result<Bytes, Error>;
 
-    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-        poll_with_read(&mut self.inner, &mut self.buf)
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut buf = self.buf.clone();
+        if let Ok(mut reader) = self.reader.lock() {
+            reader.waker = Some(cx.waker().clone());
+        }
+        poll_with_read(&mut self.inner, &mut buf)
     }
 }
 
 /// A deflate decoder that reads from a `deflate::Decoder` into a `BytesMut` and emits the results
-/// as a `Chunk`.
+/// as a `Bytes`.
 struct Deflate {
     inner: Box<DeflateDecoder<Peeked<ReadableChunks<Body>>>>,
     buf: BytesMut,
+    reader: Arc<Mutex<ReadableChunks<Body>>>,
 }
 
 impl Deflate {
     fn new(stream: ReadableChunks<Body>) -> Self {
+        let stream = Arc::new(Mutex::new(stream));
+        let reader = stream.clone();
         Self {
             buf: BytesMut::with_capacity(INIT_BUFFER_SIZE),
             inner: Box::new(DeflateDecoder::new(Peeked::new(stream))),
+            reader: reader,
         }
     }
 }
 
 impl Stream for Deflate {
-    type Item = Chunk;
-    type Error = Error;
+    type Item = Result<Bytes, Error>;
 
-    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-        poll_with_read(&mut self.inner, &mut self.buf)
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut buf = self.buf.clone();
+        if let Ok(mut reader) = self.reader.lock() {
+            reader.waker = Some(cx.waker().clone());
```

Comment:
necessary, since we need to ensure the underlying ReadableChunks Stream has access to the passed in waker via the Context. This is the recommended way to do it:

https://rust-lang.github.io/async-book/02_execution/03_wakeups.html